### PR TITLE
evm: implement vm.GetHashFn

### LINKED ISSUE #618: In the "newEVM" function, the member variable "GetHash" is not initialized when "vm.Context" is initialized
__System info:__ 
* in branch "development"

__Expected behavior:__ [What you expected to happen]
* function newEVM
```golang

func (st StateTransition) newEVM(ctx sdk.Context, csdb *CommitStateDB, gasLimit uint64, gasPrice *big.Int, config ChainConfig) *vm.EVM {
	// Create context for evm
	context := vm.Context{
		CanTransfer: core.CanTransfer,
		Transfer:    core.Transfer,
                 GetHash: FUNCTION_GET_HASH,
		Origin:      st.Sender,
		Coinbase:    common.Address{}, // there's no benefitiary since we're not mining
		BlockNumber: big.NewInt(ctx.BlockHeight()),
		Time:        big.NewInt(ctx.BlockHeader().Time.Unix()),
		Difficulty:  big.NewInt(0), // unused. Only required in PoW context
		GasLimit:    gasLimit,
		GasPrice:    gasPrice,
	}

	return vm.NewEVM(context, csdb, config.EthereumConfig(st.ChainID), vm.Config{})
}
```
__Actual behavior:__ [What actually happened]
* function newEVM
```golang

func (st StateTransition) newEVM(ctx sdk.Context, csdb *CommitStateDB, gasLimit uint64, gasPrice *big.Int, config ChainConfig) *vm.EVM {
	// Create context for evm
	context := vm.Context{
		CanTransfer: core.CanTransfer,
		Transfer:    core.Transfer,
		Origin:      st.Sender,
		Coinbase:    common.Address{}, // there's no benefitiary since we're not mining
		BlockNumber: big.NewInt(ctx.BlockHeight()),
		Time:        big.NewInt(ctx.BlockHeader().Time.Unix()),
		Difficulty:  big.NewInt(0), // unused. Only required in PoW context
		GasLimit:    gasLimit,
		GasPrice:    gasPrice,
	}

	return vm.NewEVM(context, csdb, config.EthereumConfig(st.ChainID), vm.Config{})
}
```
__Additional info:__ [Include gist of relevant config, logs, etc.]
* the struct "vm.Context" have a member variable called "GetHash"
```golang

// Context provides the EVM with auxiliary information. Once provided
// it shouldn't be modified.
type Context struct {
	// CanTransfer returns whether the account contains
	// sufficient ether to transfer the value
	CanTransfer CanTransferFunc
	// Transfer transfers ether from one account to the other
	Transfer TransferFunc
	// GetHash returns the hash corresponding to n
	GetHash GetHashFunc

	// Message information
	Origin   common.Address // Provides information for ORIGIN
	GasPrice *big.Int       // Provides information for GASPRICE

	// Block information
	Coinbase    common.Address // Provides information for COINBASE
	GasLimit    uint64         // Provides information for GASLIMIT
	BlockNumber *big.Int       // Provides information for NUMBER
	Time        *big.Int       // Provides information for TIME
	Difficulty  *big.Int       // Provides information for DIFFICULTY
}
```

* function "GetHash" from "vm.Context" is used in opCode of evm
```golang

func opBlockhash(pc *uint64, interpreter *EVMInterpreter, callContext *callCtx) ([]byte, error) {
	num := callContext.stack.peek()
	num64, overflow := num.Uint64WithOverflow()
	if overflow {
		num.Clear()
		return nil, nil
	}
	var upper, lower uint64
	upper = interpreter.evm.BlockNumber.Uint64()
	if upper < 257 {
		lower = 0
	} else {
		lower = upper - 256
	}
	if num64 >= lower && num64 < upper {
		num.SetBytes(interpreter.evm.GetHash(num64).Bytes())
	} else {
		num.Clear()
	}
	return nil, nil
}
```
* If the method opBlockhash is called when the contract is executed, it may panic because GetHash is nil

### PR DESCRIPTION
**Pull request by fedekunze**
closes #618  

TODO:

- [x] tests
- [x] store hash on end block

### CODE REVIEW

**Comment by fedekunze** [comment]
actually, we will need to store the (height, hash) pair on the EVM state

File: x/evm/keeper/abci.go [hunk 1]
```diff
@@ -29,18 +32,31 @@ func (k *Keeper) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
...
+	if err != nil {
+		panic(err)
+	}
+
+	// set the hash for the current height and epoch
+	// NOTE: we set the hash here instead of on BeginBlock in order to set the final block prior to
+	// an upgrade. If we set it on BeginBlock the last block from prior to the upgrade wouldn't be
+	// included on the store.
+	hash := types.HashFromContext(ctx)
+	k.SetHeightHash(ctx, epoch.Uint64(), uint64(ctx.BlockHeight()), hash)
```
Comment (summerpro) [inline x/evm/keeper/abci.go]: * May I ask why we use "epoch" as the key here? Is it possible to just use blockHeight?

**Review by noot — APPROVED** [review]
looks good, also wondering about the epoch, is there a case where the different epochs/chainIDs would be using the same store?

**Comment by fedekunze** [comment]
> looks good, also wondering about the epoch, is there a case where the different epochs/chainIDs would be using the same store?

Yeah so after an upgrade the chain height could be reset and thus a new height from the latest epoch can override the previous one. After giving it a second thought, I think that could be ok as we care for the latest height only.

**Comment by fedekunze** [comment]
updated @noot

**Review by araskachoi — COMMENTED** [review]
Question: why don't we use `csdb.WithContext(ctx).GetHeightHash(height)` here as well?

File: x/evm/types/state_transition.go [hunk 2]
```diff
@@ -47,11 +47,42 @@ type ExecutionResult struct {
...
+//  1. The requested height matches the current height from context (and thus same epoch number)
+//  2. The requested height is from an previous height from the same chain epoch
+//  3. The requested height is from a height greater than the latest one
+func GetHashFn(ctx sdk.Context, csdb *CommitStateDB) vm.GetHashFunc {
+	return func(height uint64) common.Hash {
+		switch {
+		case ctx.BlockHeight() == int64(height):
+			// Case 1: The requested height matches the one from the context so we can retrieve the header
+			// hash directly from the context.
+			return HashFromContext(ctx)
```
Comment (araskachoi) [inline x/evm/types/state_transition.go:60]: Question: why don't we use `csdb.WithContext(ctx).GetHeightHash(height)` here as well?

**Pull request merged**